In [1]:
import numpy as np
from keras.datasets import fashion_mnist as fmn
from keras.models import Sequential
from keras.models import Model
from keras.layers import Input, Convolution1D, MaxPooling1D, Dense, Dropout, Flatten
from keras.utils import to_categorical
from keras.callbacks import EarlyStopping

/usr/local/anaconda3/lib/python3.6/site-packages/h5py/__init__.py:34: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
Using TensorFlow backend.


# Problem 1
Build a CNN comparable in structure to the one in the reading you did on CNNs, but train and test it on the fashion-MNIST dataset.  Adjust some of the parameters and compare the results.  You should be able to get performance better than any of the classifiers we have used on Fashion-MNIST so far.

In [2]:
# Load the data.
(train_x,train_y),(test_x,test_y) = fmn.load_data()

# Get important nums about data
num_train, height, width = train_x.shape
num_test = test_x.shape[0]
num_classes = np.unique(train_y).shape[0]

# Standardize data
train_x = train_x.astype('float32')
test_x = test_x.astype('float32')
train_x /= np.max(train_x)
test_x /= np.max(test_x)

train_y = to_categorical(train_y,num_classes)
test_y = to_categorical(test_y,num_classes)

In [3]:
# Set up constants
batch_size = 32
num_epochs = 200
kernel_size = 3
pool_size = 2
conv_depth_1 = 32
conv_depth_2 = 64
drop_prob_1 = 0.25
drop_prob_2 = 0.5
hidden_size = 512

In [4]:
# Build your CNN.
inputs = Input(shape=(height, width))
conv_1 = Convolution1D(conv_depth_1,kernel_size,padding='same',activation='relu')(inputs)
conv_2 = Convolution1D(conv_depth_1,kernel_size,padding='same',activation='relu')(conv_1)
pool_1 = MaxPooling1D(pool_size=pool_size)(conv_2)
drop_1 = Dropout(drop_prob_1)(pool_1)
conv_3 = Convolution1D(conv_depth_2,kernel_size,padding='same',activation='relu')(drop_1)
conv_4 = Convolution1D(conv_depth_2,kernel_size,padding='same',activation='relu')(conv_3)
pool_2 = MaxPooling1D(pool_size=pool_size)(conv_4)
drop_2 = Dropout(drop_prob_1)(pool_2)
flat = Flatten()(drop_2)
hidden = Dense(hidden_size, activation='relu')(flat)
drop_3 = Dropout(drop_prob_2)(hidden)
out = Dense(num_classes,activation='softmax')(drop_3)

model = Model(inputs=inputs, outputs=out)
model.compile(loss='categorical_crossentropy',optimizer='adam',metrics=['accuracy'])
early_stopping = EarlyStopping(monitor='acc',patience=2)

In [5]:
# Train and test.
model.fit(train_x,train_y,batch_size=batch_size,epochs=num_epochs,verbose=1,callbacks=[early_stopping])
model.evaluate(test_x,test_y,verbose=1)

Epoch 1/200
60000/60000 [==============================] - 11s 191us/step - loss: 0.5761 - acc: 0.7893
Epoch 2/200
60000/60000 [==============================] - 11s 189us/step - loss: 0.4062 - acc: 0.8507
Epoch 3/200
60000/60000 [==============================] - 11s 186us/step - loss: 0.3719 - acc: 0.8625
Epoch 4/200
60000/60000 [==============================] - 11s 177us/step - loss: 0.3514 - acc: 0.8689
Epoch 5/200
60000/60000 [==============================] - 10s 175us/step - loss: 0.3368 - acc: 0.8753
Epoch 6/200
60000/60000 [==============================] - 11s 184us/step - loss: 0.3280 - acc: 0.8787
Epoch 7/200
60000/60000 [==============================] - 10s 173us/step - loss: 0.3189 - acc: 0.8821
Epoch 8/200
60000/60000 [==============================] - 11s 179us/step - loss: 0.3095 - acc: 0.8837
Epoch 9/200
60000/60000 [==============================] - 11s 183us/step - loss: 0.3067 - acc: 0.8867
Epoch 10/200
60000/60000 [==============================] - 11s 179us/ste

[0.28503741631507873, 0.896]

In [6]:
# Adjust parameters.
# Set up constants
batch_size = 32
num_epochs = 200
kernel_size = 5
pool_size = 3
conv_depth_1 = 32
conv_depth_2 = 64
drop_prob_1 = 0.25
drop_prob_2 = 0.35
hidden_size = 1024
# Build your CNN.
inputs = Input(shape=(height, width))
conv_1 = Convolution1D(conv_depth_1,kernel_size,padding='same',activation='relu')(inputs)
conv_2 = Convolution1D(conv_depth_1,kernel_size,padding='same',activation='relu')(conv_1)
pool_1 = MaxPooling1D(pool_size=pool_size)(conv_2)
drop_1 = Dropout(drop_prob_1)(pool_1)
conv_3 = Convolution1D(conv_depth_2,kernel_size,padding='same',activation='relu')(drop_1)
conv_4 = Convolution1D(conv_depth_2,kernel_size,padding='same',activation='relu')(conv_3)
pool_2 = MaxPooling1D(pool_size=pool_size)(conv_4)
drop_2 = Dropout(drop_prob_1)(pool_2)
flat = Flatten()(drop_2)
hidden = Dense(hidden_size, activation='relu')(flat)
drop_3 = Dropout(drop_prob_2)(hidden)
out = Dense(num_classes,activation='softmax')(drop_3)

model = Model(inputs=inputs, outputs=out)
model.compile(loss='categorical_crossentropy',optimizer='adam',metrics=['accuracy'])
early_stopping = EarlyStopping(monitor='acc',patience=2)
# Train and test.
model.fit(train_x,train_y,batch_size=batch_size,epochs=num_epochs,verbose=1,callbacks=[early_stopping])
model.evaluate(test_x,test_y,verbose=1)

Epoch 1/200
60000/60000 [==============================] - 12s 196us/step - loss: 0.5668 - acc: 0.7893
Epoch 2/200
60000/60000 [==============================] - 12s 196us/step - loss: 0.4110 - acc: 0.8474
Epoch 3/200
60000/60000 [==============================] - 12s 194us/step - loss: 0.3761 - acc: 0.8593
Epoch 4/200
60000/60000 [==============================] - 11s 185us/step - loss: 0.3518 - acc: 0.8689
Epoch 5/200
60000/60000 [==============================] - 12s 196us/step - loss: 0.3383 - acc: 0.8746
Epoch 6/200
60000/60000 [==============================] - 11s 188us/step - loss: 0.3280 - acc: 0.8770
Epoch 7/200
60000/60000 [==============================] - 12s 196us/step - loss: 0.3212 - acc: 0.8798
Epoch 8/200
60000/60000 [==============================] - 13s 223us/step - loss: 0.3136 - acc: 0.8837
Epoch 9/200
60000/60000 [==============================] - 14s 227us/step - loss: 0.3075 - acc: 0.8849
Epoch 10/200
60000/60000 [==============================] - 13s 221us/ste

[0.3066774897933006, 0.8856]

In [7]:
# Adjust parameters.
# Set up constants
batch_size = 32
num_epochs = 200
kernel_size = 2
pool_size = 1
conv_depth_1 = 16
conv_depth_2 = 64
drop_prob_1 = 0.15
drop_prob_2 = 0.35
hidden_size = 1024
# Build your CNN.
inputs = Input(shape=(height, width))
conv_1 = Convolution1D(conv_depth_1,kernel_size,padding='same',activation='relu')(inputs)
conv_2 = Convolution1D(conv_depth_1,kernel_size,padding='same',activation='relu')(conv_1)
pool_1 = MaxPooling1D(pool_size=pool_size)(conv_2)
drop_1 = Dropout(drop_prob_1)(pool_1)
conv_3 = Convolution1D(conv_depth_2,kernel_size,padding='same',activation='relu')(drop_1)
conv_4 = Convolution1D(conv_depth_2,kernel_size,padding='same',activation='relu')(conv_3)
pool_2 = MaxPooling1D(pool_size=pool_size)(conv_4)
drop_2 = Dropout(drop_prob_1)(pool_2)
flat = Flatten()(drop_2)
hidden = Dense(hidden_size, activation='relu')(flat)
drop_3 = Dropout(drop_prob_2)(hidden)
out = Dense(num_classes,activation='softmax')(drop_3)

model = Model(inputs=inputs, outputs=out)
model.compile(loss='categorical_crossentropy',optimizer='adam',metrics=['accuracy'])
early_stopping = EarlyStopping(monitor='acc',patience=2)
# Train and test.
model.fit(train_x,train_y,batch_size=batch_size,epochs=num_epochs,verbose=1,callbacks=[early_stopping])
model.evaluate(test_x,test_y,verbose=1)

Epoch 1/200
60000/60000 [==============================] - 51s 842us/step - loss: 0.5132 - acc: 0.8118
Epoch 2/200
60000/60000 [==============================] - 46s 767us/step - loss: 0.3749 - acc: 0.8610
Epoch 3/200
60000/60000 [==============================] - 46s 769us/step - loss: 0.3392 - acc: 0.8739
Epoch 4/200
60000/60000 [==============================] - 47s 783us/step - loss: 0.3159 - acc: 0.8819
Epoch 5/200
60000/60000 [==============================] - 47s 780us/step - loss: 0.2973 - acc: 0.8886
Epoch 6/200
60000/60000 [==============================] - 46s 759us/step - loss: 0.2826 - acc: 0.8939
Epoch 7/200
60000/60000 [==============================] - 50s 836us/step - loss: 0.2670 - acc: 0.8988
Epoch 8/200
60000/60000 [==============================] - 48s 803us/step - loss: 0.2574 - acc: 0.9029
Epoch 9/200
60000/60000 [==============================] - 44s 736us/step - loss: 0.2453 - acc: 0.9079
Epoch 10/200
60000/60000 [==============================] - 44s 738us/ste

[0.4067879297107458, 0.8965]

In [8]:
# Adjust parameters.
# Set up constants
batch_size = 32
num_epochs = 200
kernel_size = 8
pool_size = 4
conv_depth_1 = 64
conv_depth_2 = 128
drop_prob_1 = 0.25
drop_prob_2 = 0.55
hidden_size = 512
# Build your CNN.
inputs = Input(shape=(height, width))
conv_1 = Convolution1D(conv_depth_1,kernel_size,padding='same',activation='relu')(inputs)
conv_2 = Convolution1D(conv_depth_1,kernel_size,padding='same',activation='relu')(conv_1)
pool_1 = MaxPooling1D(pool_size=pool_size)(conv_2)
drop_1 = Dropout(drop_prob_1)(pool_1)
conv_3 = Convolution1D(conv_depth_2,kernel_size,padding='same',activation='relu')(drop_1)
conv_4 = Convolution1D(conv_depth_2,kernel_size,padding='same',activation='relu')(conv_3)
pool_2 = MaxPooling1D(pool_size=pool_size)(conv_4)
drop_2 = Dropout(drop_prob_1)(pool_2)
flat = Flatten()(drop_2)
hidden = Dense(hidden_size, activation='relu')(flat)
drop_3 = Dropout(drop_prob_2)(hidden)
out = Dense(num_classes,activation='softmax')(drop_3)

model = Model(inputs=inputs, outputs=out)
model.compile(loss='categorical_crossentropy',optimizer='adam',metrics=['accuracy'])
early_stopping = EarlyStopping(monitor='acc',patience=2)
# Train and test.
model.fit(train_x,train_y,batch_size=batch_size,epochs=num_epochs,verbose=1,callbacks=[early_stopping])
model.evaluate(test_x,test_y,verbose=1)

Epoch 1/200
60000/60000 [==============================] - 23s 376us/step - loss: 0.5725 - acc: 0.7906
Epoch 2/200
60000/60000 [==============================] - 22s 371us/step - loss: 0.3959 - acc: 0.8585
Epoch 3/200
60000/60000 [==============================] - 22s 365us/step - loss: 0.3563 - acc: 0.8721
Epoch 4/200
60000/60000 [==============================] - 22s 369us/step - loss: 0.3352 - acc: 0.8791
Epoch 5/200
60000/60000 [==============================] - 23s 378us/step - loss: 0.3195 - acc: 0.8842
Epoch 6/200
60000/60000 [==============================] - 23s 379us/step - loss: 0.3043 - acc: 0.8892
Epoch 7/200
60000/60000 [==============================] - 23s 378us/step - loss: 0.2974 - acc: 0.8919
Epoch 8/200
60000/60000 [==============================] - 23s 378us/step - loss: 0.2931 - acc: 0.8937
Epoch 9/200
60000/60000 [==============================] - 22s 371us/step - loss: 0.2831 - acc: 0.8974
Epoch 10/200
60000/60000 [==============================] - 22s 372us/ste

[0.3061091561436653, 0.8971]

Compare Results :

The original results were the best, at 89.6% accuracy taking 22 epochs of about 11 seconds a piece.  When I increased the kernel and pool size as well as the size of the hidden layer, it stopped after only 15 epochs of about 13 seconds a piece and lost one percent accuracy.  Decreasing the kernel and pool size had the unfortunate effect of overfitting a little, since the training data measured up to 96% accuracy, but when tested on the test data, it only managed 89.65%.  While this is greater than the original accuracy, it not only isn't a very big improvement, but it required 59 epochs of about 45 seconds a piece, far longer than the original parameters.  This indicates to me that I shouldn't have used training accuracy as my indicator to stop training, but rather a validation of the test data.  The last set seems to have performed the second best, taking about twice as long as the original set and giving a better accuracy (one of 89.71%).  These took about 23 seconds a piece, and finished in just 24 epochs.